modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotCheckProcess.java

```diff
@@ -0,0 +1,740 @@
+        }
+
+        return phaseFut;
+    }
+
+    /** Phase 1 end. */
+    private void reducePreparationAndMetasCheck(
+        UUID reqId,
+        Map<UUID, SnapshotCheckResponse> results,
+        Map<UUID, Throwable> errors
+    ) {
+        SnapshotCheckContext ctx = context(null, reqId);
+
+        // The context is not stored in the case of concurrent check of the same snapshot but the operation future is registered.
+        GridFutureAdapter<SnapshotPartitionsVerifyTaskResult> clusterOpFut = clusterOpFuts.get(reqId);
+
+        try {
+            if (!errors.isEmpty())
+                throw new IgniteSnapshotVerifyException(mapErrors(errors));
+
+            if (ctx == null) {
+                assert clusterOpFut == null;
+
+                return;
+            }
+
+            ClusterTopologyCheckedException ex;
+
+            if ((ex = checkNodeLeft(ctx.req.nodes(), results.keySet())) != null)
+                throw ex;
+
+            Map<ClusterNode, List<SnapshotMetadata>> metas = new HashMap<>();
+
+            results.forEach((nodeId, nodeRes) -> {
+                // A node might be not required. It gives null result. But a required node might have invalid empty result
+                // which must be validated.
+                if (ctx.req.nodes().contains(nodeId) && baseline(nodeId) && !F.isEmpty((Collection<?>)nodeRes.result())) {
+                    assert nodeRes != null;
+
+                    metas.put(kctx.cluster().get().node(nodeId), nodeRes.result());
+                }
+            });
+
+            Map<ClusterNode, Exception> metasCheck = snpChecker.reduceMetasResults(
+                new SnapshotFileTree(kctx, ctx.req.snapshotName(), ctx.req.snapshotPath()), metas);
+
+            if (!metasCheck.isEmpty())
+                throw new IgniteSnapshotVerifyException(metasCheck);
+
+            // If the topology is lesser that the snapshot's, we have to check another partitions parts.
+            ctx.metas = assingMetas(metas);
+
+            if (!F.isEmpty(ctx.metas)) {
+                ctx.locFileTree = new HashMap<>(ctx.metas.size(), 1.0f);
+
+                for (SnapshotMetadata metaToProc : ctx.metas) {
+                    SnapshotFileTree sft = new SnapshotFileTree(kctx, ctx.req.snapshotName(), ctx.req.snapshotPath(),
+                        metaToProc.folderName(), metaToProc.consistentId());
+
+                    ctx.locFileTree.put(metaToProc.consistentId(), sft);
+                }
+            }
+
+            if (clusterOpFut != null)
+                ctx.clusterMetas = metas;
+
+            if (U.isLocalNodeCoordinator(kctx.discovery()))
+                phase2PartsHashes.start(reqId, ctx.req);
+        }
+        catch (Throwable th) {
+            if (ctx != null) {
+                contexts.remove(ctx.req.snapshotName());
+
+                if (log.isInfoEnabled())
+                    log.info("Finished snapshot validation with error [req=" + ctx.req + ", err=" + th.getMessage() + ']');
```

Comment:
We have `Finished snapshot validation [req=` in the other plases. I would't want to parse additional `with error`. `, err=` looks enough.